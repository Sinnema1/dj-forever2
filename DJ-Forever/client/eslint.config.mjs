import js from '@eslint/js';
import tseslint from 'typescript-eslint';
import react from 'eslint-plugin-react';
import reactHooks from 'eslint-plugin-react-hooks';
import prettier from 'eslint-plugin-prettier';

export default /** @type {import('eslint').Linter.FlatConfig[]} */ ([
  {
    ignores: ['dist/', 'build/', 'node_modules/', 'vite.config.ts'],
  },
  js.configs.recommended,
  ...tseslint.configs.recommended,
  {
    files: ['**/*.ts', '**/*.tsx'],
    languageOptions: {
      parser: tseslint.parser,
      parserOptions: {
        project: './tsconfig.json',
        sourceType: 'module',
        ecmaVersion: 2021,
        ecmaFeatures: { jsx: true },
      },
    },
    plugins: {
      '@typescript-eslint': tseslint.plugin,
      react,
      'react-hooks': reactHooks,
      prettier,
    },
    settings: {
      react: {
        version: 'detect',
      },
    },
    rules: {
      'prettier/prettier': 'warn',
      'react/react-in-jsx-scope': 'off',
      'react/prop-types': 'off',
      '@typescript-eslint/explicit-module-boundary-types': 'off',
      '@typescript-eslint/no-unused-vars': ['warn'],
      'react-hooks/rules-of-hooks': 'error',
      'react-hooks/exhaustive-deps': 'warn',
      'no-console': 'warn',
      'no-debugger': 'warn',
      'no-unused-vars': 'off',
      'import/order': [
        'warn',
        {
          groups: ['builtin', 'external', 'internal', 'parent', 'sibling', 'index'],
          'newlines-between': 'always',
        },
      ],
    },
  },
]);
