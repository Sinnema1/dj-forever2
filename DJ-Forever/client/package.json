{
  "name": "client",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.13.1",
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0",
    "@mui/icons-material": "^6.4.7",
    "@mui/material": "^6.4.7",
    "@mui/system": "^6.4.7",
    "@testing-library/dom": "^10.4.0",
    "@types/react": "^19.0.10",
    "@types/react-dom": "^19.0.4",
    "axios": "^1.8.4",
    "graphql": "^16.10.0",
    "jwt-decode": "^4.0.0",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-router-dom": "^7.3.0",
    "typescript": "^4.9.5",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "dev": "vite",
    "build": "tsc -b && vite build",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "eslint . --ext .ts,.tsx --fix",
    "preview": "vite preview",
    "test": "vitest --environment jsdom",
    "test:ui": "vitest --ui",
    "test:coverage": "vitest run --coverage"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@eslint/js": "^9.22.0",
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.3.0",
    "@testing-library/user-event": "^14.6.1",
    "@types/jwt-decode": "^2.2.1",
    "@types/node": "^18.19.110",
    "@types/react-router-dom": "^5.3.3",
    "@typescript-eslint/eslint-plugin": "^8.26.1",
    "@typescript-eslint/parser": "^8.26.1",
    "@vitejs/plugin-react": "^4.5.0",
    "eslint": "^9.22.0",
    "eslint-config-prettier": "^10.1.1",
    "eslint-plugin-prettier": "^5.2.3",
    "eslint-plugin-react": "^7.37.4",
    "eslint-plugin-react-hooks": "^5.2.0",
    "jsdom": "^26.1.0",
    "msw": "^2.8.7",
    "prettier": "^3.5.3",
    "rollup-plugin-visualizer": "^6.0.1",
    "typescript-eslint": "^8.26.1",
    "vitest": "^3.1.4"
  }
}
