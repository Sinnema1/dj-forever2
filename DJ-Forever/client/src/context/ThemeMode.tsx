import React, { createContext, useMemo, useState, ReactNode } from 'react';
import { ThemeProvider, CssBaseline } from '@mui/material';
import { getDesignTokens } from '../styles/theme';
import { createTheme } from '@mui/material/styles';

interface ThemeModeContextType {
  toggleColorMode: () => void;
  mode: 'light' | 'dark';
}

export const ColorModeContext = createContext<ThemeModeContextType>({
  toggleColorMode: () => {},
  mode: 'light',
});

export const ThemeModeProvider = ({ children }: { children: ReactNode }) => {
  const [mode, setMode] = useState<'light' | 'dark'>('light');

  const colorMode = useMemo(
    () => ({
      toggleColorMode: () => setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light')),
      mode,
    }),
    [mode],
  );

  const theme = useMemo(() => createTheme(getDesignTokens(mode)), [mode]);

  return (
    <ColorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </ColorModeContext.Provider>
  );
};
