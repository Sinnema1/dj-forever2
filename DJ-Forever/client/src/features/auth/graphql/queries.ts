import { gql } from '@apollo/client';

/**
 * Query to fetch the current logged-in user's profile.
 */
export const GET_ME = gql`
  query GetMe {
    me {
      _id
      email
      fullName
      isAdmin
    }
  }
`;
