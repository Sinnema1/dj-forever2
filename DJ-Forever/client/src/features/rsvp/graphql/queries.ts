import { gql } from '@apollo/client';

/**
 * Get all RSVPs (Admin access or general list)
 */
export const GET_RSVPS = gql`
  query GetRSVPs {
    getRSVPs {
      _id
      userId
      attending
      mealPreference
      allergies
      additionalNotes
      fullName
    }
  }
`;

/**
 * Get a single RSVP by ID (if you need it somewhere)
 */
export const GET_RSVP_BY_ID = gql`
  query GetRSVPById($id: ID!) {
    getRSVPById(id: $id) {
      _id
      userId
      attending
      mealPreference
      allergies
      additionalNotes
      fullName
    }
  }
`;

/**
 * Get the current user's RSVP (if authenticated)
 */
export const GET_RSVP = gql`
  query GetRSVP {
    getRSVP {
      _id
      userId
      attending
      mealPreference
      allergies
      additionalNotes
      fullName
    }
  }
`;
