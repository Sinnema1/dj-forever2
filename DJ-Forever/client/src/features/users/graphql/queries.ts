import { gql } from '@apollo/client';

export const GET_USERS = gql`
  query GetUsers {
    getUsers {
      _id
      fullName
      email
      isAdmin
    }
  }
`;

export const GET_USER_BY_ID = gql`
  query GetUserById($id: ID!) {
    getUserById(id: $id) {
      _id
      fullName
      email
      isAdmin
    }
  }
`;

export const GET_ME = gql`
  query Me {
    me {
      _id
      fullName
      email
      isAdmin
      hasRSVPed
      rsvpId
      isInvited
    }
  }
`;
