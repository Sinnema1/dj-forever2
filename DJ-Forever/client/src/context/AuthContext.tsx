import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { useMutation } from '@apollo/client';
import { LOGIN_USER, REGISTER_USER } from '../features/auth/graphql/mutations';
import { AuthContextType, UserType } from '../features/users/types/userTypes';

// Create context with the AuthContextType
export const AuthContext = createContext<AuthContextType | undefined>(undefined);

// AuthProvider component
export const AuthProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<UserType | null>(null);
  const [token, setToken] = useState<string | null>(null);

  // Apollo Mutations for login and register
  const [loginMutation] = useMutation(LOGIN_USER);
  const [registerMutation] = useMutation(REGISTER_USER);

  // On component mount: load token and user from localStorage (if available)
  useEffect(() => {
    const storedToken = localStorage.getItem('id_token');
    const storedUser = localStorage.getItem('user');

    if (storedToken && storedUser) {
      setToken(storedToken);
      setUser(JSON.parse(storedUser));
    }
  }, []);

  /**
   * Login Function
   * @param email User's email
   * @param password User's password
   */
  const login = async (email: string, password: string) => {
    try {
      const { data } = await loginMutation({
        variables: { email, password },
      });

      const authToken: string = data?.loginUser?.token;
      const authUser: UserType = data?.loginUser?.user;

      if (!authToken || !authUser) {
        throw new Error('Invalid login response.');
      }

      // Optional: warn if isInvited is missing from the response
      if (authUser.isInvited === undefined) {
        console.warn('Warning: isInvited field is missing from the login response.');
      }

      // Update state and localStorage
      setToken(authToken);
      setUser(authUser);
      localStorage.setItem('id_token', authToken);
      localStorage.setItem('user', JSON.stringify(authUser));
    } catch (error: unknown) {
      if (error instanceof Error) {
        throw new Error(error.message || 'Login failed.');
      }
      throw new Error('Login failed.');
    }
  };

  /**
   * Register Function
   * @param fullName User's full name
   * @param email User's email
   * @param password User's password
   */
  const registerUser = async (fullName: string, email: string, password: string) => {
    try {
      const { data } = await registerMutation({
        variables: { fullName, email, password },
      });

      // Use the correct response key from the GraphQL mutation
      const authToken: string = data?.registerUser?.token;
      const authUser: UserType = data?.registerUser?.user;

      if (!authToken || !authUser) {
        throw new Error('Invalid registration response.');
      }

      // Warn if isInvited is true for a self-registered user
      if (authUser.isInvited === true) {
        console.warn('Warning: Self-registered user should not have isInvited=true.');
      }
      if (authUser.isInvited === undefined) {
        console.warn('Warning: isInvited field is missing from the registration response.');
      }

      setToken(authToken);
      setUser(authUser);
      localStorage.setItem('id_token', authToken);
      localStorage.setItem('user', JSON.stringify(authUser));
    } catch (error: unknown) {
      if (error instanceof Error) {
        throw new Error(error.message || 'Registration failed.');
      }
      throw new Error('Registration failed.');
    }
  };

  /**
   * Logout Function
   */
  const logout = () => {
    setToken(null);
    setUser(null);
    localStorage.removeItem('id_token');
    localStorage.removeItem('user');
  };

  /**
   * Update user information in context and localStorage
   */
  const updateUserInfo = (userData: Partial<UserType>) => {
    if (!user) return;

    const updatedUser = { ...user, ...userData };
    setUser(updatedUser);
    localStorage.setItem('user', JSON.stringify(updatedUser));
  };

  // Context value exposed to consumers
  const value: AuthContextType = {
    user,
    isLoggedIn: !!token,
    login,
    registerUser,
    logout,
    updateUserInfo,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

/**
 * useAuth hook for consuming AuthContext
 * Throws an error if used outside the AuthProvider
 */
export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) throw new Error('useAuth must be used within an AuthProvider');
  return context;
};
