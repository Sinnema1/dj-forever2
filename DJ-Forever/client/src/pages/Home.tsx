import React from 'react';
import { Box, Container, Typography, Button } from '@mui/material';
import { useNavigate } from 'react-router-dom';
import { useAuth } from '../context/AuthContext';
import { useRSVP } from '../features/rsvp/hooks/useRSVP';

const Home: React.FC = () => {
  const navigate = useNavigate();
  const { user, isLoggedIn } = useAuth();
  const { rsvp } = useRSVP();

  // Find if the logged-in user has RSVP'd
  const hasRSVP = isLoggedIn && !!rsvp;

  const handleGetStarted = () => {
    if (!isLoggedIn) {
      navigate('/login');
    } else if (!hasRSVP) {
      navigate('/register');
    } else {
      navigate('/dashboard');
    }
  };

  return (
    <Box
      sx={{
        py: { xs: 4, md: 6 }, // Reduce top/bottom padding
        flex: 1,
        minHeight: '100%',
        background: (theme) =>
          `linear-gradient(135deg, ${theme.palette.primary.main} 30%, ${theme.palette.primary.light} 90%)`,
        color: '#fff',
        display: 'flex',
        alignItems: 'center',
      }}
    >
      <Container maxWidth="md" sx={{ textAlign: 'center' }}>
        <Typography variant="h2" component="h1" gutterBottom sx={{ fontWeight: 'bold' }}>
          Welcome to DJ Forever
        </Typography>
        <Typography variant="h5" gutterBottom>
          Your Ultimate Event RSVP Management Platform
        </Typography>
        <Button variant="contained" color="secondary" sx={{ mt: 4 }} onClick={handleGetStarted}>
          Get Started
        </Button>
      </Container>
    </Box>
  );
};

export default Home;
