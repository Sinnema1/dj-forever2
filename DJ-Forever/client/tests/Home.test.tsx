import { describe, it, expect } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Home from '../src/pages/Home';
import { MemoryRouter } from 'react-router-dom';
import { AuthProvider } from '../src/context/AuthContext';
import { MockedProvider } from '@apollo/client/testing';
import { GET_RSVP } from '../src/features/rsvp/graphql/queries';

const baseMocks = [
  {
    request: { query: GET_RSVP },
    result: { data: { getRSVP: null } },
  },
];

function renderHome({ mocks = baseMocks, route = '/' }: { mocks?: any; route?: string } = {}) {
  window.history.pushState({}, 'Test page', route);
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AuthProvider>
        <MemoryRouter initialEntries={[route]}>
          <Home />
        </MemoryRouter>
      </AuthProvider>
    </MockedProvider>,
  );
}

describe('Home page', () => {
  it('renders welcome text and Get Started button', async () => {
    renderHome();
    await waitFor(() => {
      expect(screen.getByText(/welcome to dj forever/i)).toBeInTheDocument();
      expect(screen.getByRole('button', { name: /get started/i })).toBeInTheDocument();
    });
  });

  it('navigates to login if not logged in and Get Started is clicked', async () => {
    renderHome();
    await waitFor(() =>
      expect(screen.getByRole('button', { name: /get started/i })).toBeInTheDocument(),
    );
    userEvent.click(screen.getByRole('button', { name: /get started/i }));
    expect(screen.getByRole('button', { name: /get started/i })).toBeInTheDocument();
  });

  it('shows correct UI if user has RSVP', async () => {
    const rsvpMocks = [
      {
        request: { query: GET_RSVP },
        result: {
          data: {
            getRSVP: {
              _id: '1',
              userId: 'u1',
              attending: true,
              mealPreference: 'Vegan',
              allergies: '',
              additionalNotes: '',
              fullName: 'Test User',
              __typename: 'RSVP',
            },
          },
        },
      },
    ];
    renderHome({ mocks: rsvpMocks });
    await waitFor(() => {
      expect(screen.getByText(/welcome to dj forever/i)).toBeInTheDocument();
      expect(screen.getByRole('button', { name: /get started/i })).toBeInTheDocument();
    });
  });

  it('matches snapshot', async () => {
    const { container } = renderHome();
    await waitFor(() => {
      expect(screen.getByText(/welcome to dj forever/i)).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });
});
