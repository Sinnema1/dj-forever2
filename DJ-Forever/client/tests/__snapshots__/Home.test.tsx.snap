// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page > matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-159c3j5"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthMd css-tsvym-MuiContainer-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-6orrbu-MuiTypography-root"
      >
        Welcome to DJ Forever
      </h1>
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1f9xtph-MuiTypography-root"
      >
        Your Ultimate Event RSVP Management Platform
      </h5>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-1cwgju-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Get Started
      </button>
    </div>
  </div>
</div>
`;
